src/ring_theory/adjoin.lean

```diff
@@ -0,0 +1,166 @@
+/-
+Copyright (c) 2019 Kenny Lau. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kenny Lau
+
+Adjoining elements to form subalgebras.
+-/
+
+import ring_theory.algebra_operations ring_theory.polynomial
+
+universes u v w
+
+open lattice submodule ring
+
+namespace algebra
+
+variables {R : Type u} {A : Type v}
+variables [comm_ring R] [comm_ring A]
+variables [algebra R A] {s t : set A}
+
+theorem subset_adjoin : s ⊆ adjoin R s :=
+set.subset.trans (set.subset_union_right _ _) subset_closure
+
+theorem adjoin_le {S : subalgebra R A} (H : s ⊆ S) : adjoin R s ≤ S :=
+closure_subset $ set.union_subset S.3 H
+
+theorem adjoin_le_iff {S : subalgebra R A} : adjoin R s ≤ S ↔ s ⊆ S:=
+⟨set.subset.trans subset_adjoin, adjoin_le⟩
+
+theorem adjoin_mono (H : s ⊆ t) : adjoin R s ≤ adjoin R t :=
+closure_subset (set.subset.trans (set.union_subset_union_right _ H) subset_closure)
+
+variables (R A)
+theorem adjoin_empty : adjoin R (∅ : set A) = ⊥ :=
+eq_bot_iff.2 $ adjoin_le $ set.empty_subset _
+variables {A}
+
+variables (s t)
+theorem adjoin_union : adjoin R (s ∪ t) = (adjoin R s).under (adjoin (adjoin R s) t) :=
+le_antisymm
+  (closure_mono $ set.union_subset
+    (set.range_subset_iff.2 $ λ r, or.inl ⟨algebra_map (adjoin R s) r, rfl⟩)
+    (set.union_subset_union_left _ $ λ x hxs, ⟨⟨_, subset_adjoin hxs⟩, rfl⟩))
+  (closure_subset $ set.union_subset
+    (set.range_subset_iff.2 $ λ x, adjoin_mono (set.subset_union_left _ _) x.2)
+    (set.subset.trans (set.subset_union_right _ _) subset_adjoin))
+
+theorem madjoin_eq_span : (adjoin R s).to_submodule = span R (monoid.closure s) :=
```

Comment:
@kckennylau Do you think it makes sense to turn `.to_submodule` into a coercion? Or is there a reason you chose not to do that?